### DISCUSSION TITLE
Proper code for triggering A2DP source with events and SD card?

### DISCUSSION THREAD

**Discussion by Buttmunky1** [pschatzmann/ESP32-A2DP#603] [Q&A] [answered]
Hi, sorry if this has been answered before, but I couldn't find an example on what the code should specifically look like when using the A2DP_Source module under triggered events with an SD card. Here's the code that I currently have - it's based on the [example provided by Phil here](https://www.pschatzmann.ch/home/2021/04/29/bluetooth-a2dp-streaming-from-files-on-a-sd-card/). My code does work with no issues, however I'm a complete beginner with C and I'm wondering if this is actually the optimal way of doing it or if I'm causing some memory issues down the road.

```
// Bluetooth Setup
#include "BluetoothA2DPSource.h"
#include <SPI.h>
#include <SD.h>

BluetoothA2DPSource a2dp_source;
File sound_file;
const char* file_name1 = "/file1.raw";
const char* file_name2 = "/file2.raw";
const char* file_name3 = "/file3.raw";
const int sd_ss_pin = 5;
const int frame_size_bytes = sizeof(int16_t) * 2;

// Buttons Setup
#include <ezButton.h>

#define BUTTON_K0 17
#define BUTTON_K1 26
#define BUTTON_K2 4

// Create ezButton object for each button
ezButton button1(BUTTON_K0);
ezButton button2(BUTTON_K1);
ezButton button3(BUTTON_K2);

// callback used by A2DP to provide the sound data
// This method is copied over completely from Phil's example
int32_t get_sound_data(Channels* data, int32_t len) {
  // the data in the file must be in int16 with 2 channels
  size_t result_len_bytes = sound_file.read((uint8_t*)data, len * frame_size_bytes);
  // result is in number of frames
  int32_t result_len = result_len_bytes / frame_size_bytes;
  ESP_LOGD("get_sound_data", "%d -> %d", len);
  return result_len;
}

void setup() {
  // Setup SD
  SD.begin(sd_ss_pin);

  // Start the bluetooth
  a2dp_source.start("MyMusic");
}

void loop() {
  button1.loop();  
  button2.loop();
  button3.loop();

  if (button1.isPressed()) {
    sound_file = SD.open(file_name1, FILE_READ);
    a2dp_source.start("MyMusic", get_sound_data);
  }

  if (button2.isPressed()) {
    sound_file = SD.open(file_name2, FILE_READ);
    a2dp_source.start("MyMusic", get_sound_data);
  }

  if (button3.isPressed()) {
    sound_file = SD.open(file_name3, FILE_READ);
    a2dp_source.start("MyMusic", get_sound_data);
}
```

Two main things I'm confused about here:

1. As I understand it, in my code _sound_file_ is a global variable, where its value is constantly being replaced each time a button is triggered. Is this the optimal way of doing this? I kept it this way because I did not want to change Phil's _get_sound_data_ method where it calls a global variable, but would an alternative way be to use an argument in the method, and would that be better or worse memory-wise? 

2. With regards to the overall structure, it seems like we're doing a lot of repeated computation to read the sound files, but is this to make up for the limited memory on the ESP32? It seems unfeasible to store the read sound data into a local variable somewhere.

3. _a2dp_source.start()_ confuses me slightly - is this method supposed to be used for both the initial connection to the bluetooth sink, and also to play the audio to it afterwards too? Or am I misusing it?

Thanks for any and all help, and most of all thank you for this incredible module!

**Comment by pschatzmann** [comment, marked as answer]
You never close any file: this will lead to some memory leaks...
I would suggest that you try to close it before you open the next file: e.g. if (sound_file) sound_file.close(); 

Instead of some button library you can also use the [functionality provided by the Audio Tools](https://github.com/pschatzmann/arduino-audio-tools/wiki/Interactive-Audio)

**Reply by Buttmunky1** [reply]
Awesome, thank you so much! Just to double check, I suppose the best place to put it would be right after a button click?

```
  if (button1.isPressed()) {
    if (sound_file)
      sound_file.close();
    sound_file = SD.open(file_name2, FILE_READ);
    a2dp_source.start("MyMusic", get_sound_data);
```

Was totally unaware of that was needed!

Also I somehow completely missed this wiki page on the Interactive Audio so I will definitely give it a look and adjust my code accordingly.
